src/internal/bolt/handshake.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (c) 2002-2020 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  BOLT_PROTOCOL_V3,
+  BOLT_PROTOCOL_V4_0,
+  BOLT_PROTOCOL_V4_1,
+  BOLT_PROTOCOL_V4_2,
+  BOLT_PROTOCOL_V4_3
+} from '../constants'
+
+import { alloc } from '../node'
+import { newError } from '../../error'
+import { memoize } from '../memoize'
+
+const BOLT_MAGIC_PREAMBLE = 0x6060b017
+
+function versionify (version) {
+  const major = Math.floor(version)
+  const minor = version * 10 - major * 10
+  return {
+    minor,
+    major
+  }
+}
+
+function createHandshakeMessage (versions) {
+  if (versions.length > 4) {
+    throw newError('It should not have more than 4 versions of the protocol')
+  }
+  const handshakeBuffer = alloc(5 * 4)
+
+  handshakeBuffer.writeInt32(BOLT_MAGIC_PREAMBLE)
+
+  versions.forEach(version => {
+    if (version instanceof Array) {
+      const { major, minor } = versionify(version[0])
+      const { minor: minMinor } = versionify(version[1])
+      const range = minor - minMinor
+      handshakeBuffer.writeInt32((range << 16) | (minor << 8) | major)
+    } else {
+      const { major, minor } = versionify(version)
+      handshakeBuffer.writeInt32((minor << 8) | major)
+    }
+  })
+
+  handshakeBuffer.reset()
+
+  return handshakeBuffer
+}
+
+function parseNegotiatedResponse (buffer) {
+  const h = [
+    buffer.readUInt8(),
+    buffer.readUInt8(),
+    buffer.readUInt8(),
+    buffer.readUInt8()
+  ]
+  if (h[0] === 0x48 && h[1] === 0x54 && h[2] === 0x54 && h[3] === 0x50) {
+    throw newError(
+      'Server responded HTTP. Make sure you are not trying to connect to the http endpoint ' +
+        '(HTTP defaults to port 7474 whereas BOLT defaults to port 7687)'
+    )
+  }
+  return Number(h[3] + '.' + h[2])
+}
+
+const createHandshakeMessageMemoized = memoize(createHandshakeMessage)
+
+/**
+ * @return {BaseBuffer}
+ * @private
+ */
+function newHandshakeBuffer () {
+  return createHandshakeMessageMemoized([
+    [BOLT_PROTOCOL_V4_3, BOLT_PROTOCOL_V4_2],
```

Comment:
Maybe simpler to not import the BOLT_xxx things and just write the major/minor versions here, we kind of assume the values of them anyway here, no need for verionify with Math.floor or memoize...

Reply:
Did you mean like a tuple, for instance, 4.3 being [4, 3]?